src/header.rs

```diff
@@ -1,13 +1,90 @@
 use core::marker::PhantomData;
+use core::fmt::{Debug, Formatter};
+use core::cmp::Ordering;
 
-#[derive(Clone, Copy, Debug)]
+/// Possible Types of a [`Tag`]. The names and values are taken from the example C code
+/// at the bottom of the Multiboot2 specification.
+#[repr(u32)]
+#[derive(Copy, Clone, Debug)]
+pub enum TagType {
+    // 0
+    End,
+    Cmdline,
+    BootLoaderName,
+    Module,
+    BasicMeminfo,
+    // 5
+    Bootdev,
+    Mmap,
+    Vbe,
+    Framebuffer,
+    ElfSections,
+    // 10
+    Apm,
+    Efi32,
+    Efi64,
+    Smbios,
+    AcpiOld,
+    // 15
+    AcpiNew,
+    Network,
+    EfiMmap,
+    EfiBs,
+    Efi32Ih,
+    // 20
+    Efi64Ih,
+    LoadBaseAddr,
+}
+
+// each compare/equal direction must be implemented manually
+impl PartialEq<u32> for TagType {
+    fn eq(&self, other: &u32) -> bool {
+        *self as u32 == *other
+    }
+}
+
+// each compare/equal direction must be implemented manually
+impl PartialEq<TagType> for u32 {
+    fn eq(&self, other: &TagType) -> bool {
+        *self == *other as u32
+    }
+}
+
+impl PartialOrd<u32> for TagType {
+    fn partial_cmp(&self, other: &u32) -> Option<Ordering> {
+        let num = *self as u32;
+        Some(
+            if num < *other {
+                Ordering::Less
+            } else if num == *other {
+                Ordering::Equal
+            } else {
+                Ordering::Greater
+            }
+        )
+    }
+}
+
+/// All tags that could passed via the Multiboot2 information structure to a payload/program/kernel.
+/// Better not confuse this with the Multiboot2 header tags. They are something different.
+#[derive(Clone, Copy)]
 #[repr(C)]
 pub struct Tag {
     pub typ: u32,
     pub size: u32,
     // tag specific fields
 }
 
+impl Debug for Tag {
+    fn fmt(&self, f: &mut Formatter<'_>) -> core::fmt::Result {
+        f.debug_struct("Tag")
+            .field("typ", &self.typ)
+            .field("typ (name)", unsafe { &(core::mem::transmute::<u32, TagType>(self.typ)) })
```

Comment:
any idea how this line can be written in a nicer way? I need to change the type from u32 to TagType